### PR TITLE
[MPS] Compute arange in double-float so its output is correctly rounded

### PR DESCRIPTION
**Pull request by BenjaminDEMAILLE**
Per `AI_POLICY.md`: the description below was drafted with an AI coding assistant (Claude Code) and reviewed by me before submission.

This is the "propose a workaround" half of the discussion on #193545. Metal has no `double`, but a kernel that needs more than float32 for an intermediate can carry one in double-float, and `arange` turned out to be a place where the missing precision is visible on completely ordinary inputs.

> ## Motivation
>
> `torch.arange(0, 100, 0.1)` on MPS disagreed with the correctly rounded result on 199 of its 1000 elements, and `arange(0, 1000, 0.001)` on 593070 of a million. The kernel evaluated `start + i * step` entirely in float32, and the host had already rounded `start` and `step` to float32 before binding them, so three separate roundings landed on a value that only has 24 bits to spare.
>
> ## Approach
>
> The intermediate is carried in double-float: a value held as an unevaluated sum of two float32s, giving roughly 48 bits of significand, which is enough that rounding it to float32 lands on the same value a double computation would. `c10/metal/double_float.h` adds the type and the error-free transformations it needs (Knuth's two-sum, Dekker's fast two-sum, and an fma-based exact product).
>
> Those identities hold only under IEEE semantics, which is what the shaders get: the metallib is built with `-fno-fast-math` (`cmake/Metal.cmake`) and runtime compilation uses `MTLMathModeSafe` unless `PYTORCH_MPS_FAST_MATH` says otherwise.
>
> The host now passes `start` and `step` as hi/lo pairs instead of pre-rounding them, and the float, half and bfloat16 instantiations move to the new kernels. The integer instantiations compute in int64 and were already exact, so they are untouched.
>
> The alternative was to match the CPU bit for bit, but the CPU is not the right target here: its vectorized path computes a chunk base in double and then fills the vector with `Vectorized<float>::arange(base, step)` in float32 (`aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp`), so it carries the same class of error. The reference this change is measured against is the correctly rounded `start + i * step`, which MPS now hits exactly and the CPU does not.
>
> Only `arange` is changed. `linspace` and `logspace` interpolate from both endpoints and have their own accuracy story, which is left alone.
>
> ## Test plan
>
> ```
> python test/test_mps.py TestMPS -k arange -v
> python test/test_mps.py -k linspace
> python test/test_mps.py -k logspace
> python test/test_mps.py -k test_range
> ```
>
> Accuracy before and after, counting elements that differ from the correctly rounded value computed in float64 (macOS 26.6.1, Apple M4 Max):
>
> ```
> python -c "
> import torch
> for a in [(0.0,100.0,0.1),(0.0,1000.0,0.001),(1e6,1e6+100,0.001),(0.0,1e5,0.3),(-50.0,50.0,0.7)]:
>     n = torch.arange(*a).numel()
>     ref = (a[0] + torch.arange(n, dtype=torch.float64) * a[2]).to(torch.float32)
>     m = torch.arange(*a, device='mps').cpu()
>     print(a, n, (m != ref).sum().item())
> "
> ```
>
> ```
>                                     before    after
> arange(0.0, 100.0, 0.1)                199        0   of       1000
> arange(0.0, 1000.0, 0.001)          593070        0   of    1000000
> arange(1e6, 1e6 + 100, 0.001)            0        0   of     100000
> arange(0.0, 1e5, 0.3)               156300        0   of     333334
> arange(-50.0, 50.0, 0.7)                75        0   of        143
> ```
>
> The "before" numbers are from a build of `main` at the branch point, not a simulation.

### CODE REVIEW

**Comment by Skylion007** [comment]
@Benjamine perhaps it would better to consider a more permanent solution like: https://github.com/pytorch/pytorch/issues/83264

**Comment by BenjaminDEMAILLE** [comment]
Drafted with Claude Code, reviewed by me before posting (per `AI_POLICY.md`):

Thanks, that is a useful pointer, though I think it is a larger and different problem. #83264 is about exposing an FP64 *dtype*, and its central point is dynamic range rather than precision. The `df32` here keeps the float32 exponent and only widens the significand to ~48 bits, so it does not address dynamic range and would not satisfy that issue.

 The problem this PR fixes is narrower: one intermediate inside a single kernel needed one more rounding's worth of significand. Routing that through a general software-float pipeline would cost far more than the three flops per element it takes now.

If a fuller emulation is ever built, the error-free transformations in `c10/metal/double_float.h` are where it would start rather than something it would replace, and I am happy to move or rename the header if you would rather it be positioned as the base of that work.

File: aten/src/ATen/native/mps/operations/RangeFactories.mm [hunk 1]
```diff
@@ -33,12 +33,22 @@ void arange_range_fill_mps(const Scalar& start, const Scalar& step, Tensor& resu
...
 
-  // Binds result and {start, step}: int64 for integer dtypes, float otherwise.
+  // The float kernels evaluate start + i * step in double-float, so each
+  // operand is passed as a hi/lo float pair rather than rounded to float here.
+  const auto split = [](double v) {
+    const auto hi = static_cast<float>(v);
+    const auto lo = std::isfinite(hi) ? static_cast<float>(v - static_cast<double>(hi)) : 0.0f;
+    return std::array<float, 2>{hi, lo};
+  };
+
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RangeFactories.mm]: Can you move split to `c10/metal/double_float.h` and guard it behind `#ifndef __METAL__` (see precedents)

File: c10/metal/double_float.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
+// Double-float arithmetic, for kernels that need more precision than float32
```
Comment (malfet) [inline c10/metal/double_float.h]:
```suggestion
// Double-float arithmetic library
```

File: c10/metal/double_float.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  df32 s = two_sum(a.hi, b.hi);
+  s.lo += a.lo + b.lo;
+  return quick_two_sum(s.hi, s.lo);
+}
+
+inline df32 df_mul(df32 a, df32 b) {
+  df32 p = two_prod(a.hi, b.hi);
+  p.lo += a.hi * b.lo + a.lo * b.hi;
+  return quick_two_sum(p.hi, p.lo);
+}
```
Comment (malfet) [inline c10/metal/double_float.h]: Why not re-use `c10::metal::mul` primitive here? (which is already used to implement multiplication for complex numbers

File: c10/metal/double_float.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+// Exact for |i| < 2^48, which covers any index a kernel can be dispatched on.
+inline df32 df_from_long(long i) {
+  const float hi = static_cast<float>(i);
+  return df32{hi, static_cast<float>(i - static_cast<long>(hi))};
+}
```
Comment (malfet) [inline c10/metal/double_float.h]: This feels more natural like a constructor with explicit keyword over the type, isn't it?

File: c10/metal/double_float.h [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#pragma once
+#include <metal_stdlib>
+
+namespace c10 {
+namespace metal {
+
+// A value held as the unevaluated sum hi + lo of two float32s, with
+// |lo| <= ulp(hi) / 2, carrying roughly 48 bits of significand. Metal Shading
+// Language has no `double`, so this is how a kernel keeps an intermediate at
+// better than float32 precision.
```
Comment (malfet) [inline c10/metal/double_float.h]: Would be nice to add a bit more info here, comparing double float precision to real float64 one

**Review by malfet — COMMENTED** [review]
I like the approach, but please document a bit more it:s precision limitation, and perhaps add both `mul` and `div` method, even if they are not used right now

File: test/test_mps.py [hunk 6]
```diff
@@ -9239,6 +9228,26 @@ def do_arange(start=1.2, end=10.3, dtype=torch.bfloat16, device='cpu'):
...
+            ref = (start + i * step).to(torch.float32).to(dtype)
+            self.assertEqual(out.cpu(), ref, atol=0, rtol=0,
+                             msg=f"arange({start}, {end}, {step}) mismatch for {dtype}")
+
+        # Non-contiguous out goes through the strided kernel
+        buf = torch.empty(2000, device='mps', dtype=dtype)
+        out = torch.arange(0.0, 100.0, 0.1, out=buf[::2])
+        i = torch.arange(out.numel(), dtype=torch.float64)
+        self.assertEqual(out.cpu(), (i * 0.1).to(torch.float32).to(dtype), atol=0, rtol=0)
+
```
Comment (malfet) [inline test/test_mps.py:9250]: Two questions: why this is MPS specific? And why couldn't it be done thru the OpInfo testing?

**Review by malfet — COMMENTED** [review]
If you can extract double_float into a separate PR, I'm more then happy to stamp it, but I'm not convinced why arange needs a higher precision?
I.e. for CUDA step is defined as
https://github.com/pytorch/pytorch/blob/58c235874aa64ef2925e63cc8161b03e84213704/aten/src/ATen/native/cuda/RangeFactories.cu#L224
And flor float it's float
https://github.com/pytorch/pytorch/blob/58c235874aa64ef2925e63cc8161b03e84213704/aten/src/ATen/AccumulateType.h#L136

**Comment by BenjaminDEMAILLE** [comment]
Will do

**Comment by BenjaminDEMAILLE** [comment]
Drafted with Claude Code, reviewed by me before posting (per `AI_POLICY.md`):

> Extracted into #193910.
>
> On arange, you are right and I withdraw the framing. `CUDA_ACC_TYPE(float, float)` means `arange_cuda_out` evaluates `xstart + xstep * ind` in float32 for a float32 output, which is the same formula MPS had. So MPS was not diverging from CUDA here; this PR would make it more precise than both CUDA and the CPU, which creates a divergence rather than removing one. The measurement I posted (199 of 1000 elements of `arange(0, 100, 0.1)` differing from the correctly rounded value) applies to the CUDA formula just as much.
>
> That makes it a cross-backend question about whether float32 `arange` should be correctly rounded, not an MPS fix, and not something to land through the MPS backend alone. Happy to close this unless you want to pursue it as a general change.
>
> On the test: it was MPS-specific only because the behaviour change was. It could not have gone through OpInfo, because OpInfo compares against the CPU, and the CPU vectorized path fills each chunk with `Vectorized<float>::arange(base, step)` in float32, so it is not correctly rounded either; comparing against it would have failed rather than expressed the property. That is why it was written as an explicit float64 reference check. If the behaviour change is dropped, the test goes with it.

**Comment by BenjaminDEMAILLE** [comment]
Drafted with Claude Code, reviewed by me before posting (per `AI_POLICY.md`):

> Closing this. I withdrew the arange framing above, and the part you were willing to stamp, the `double_float` type itself, is now #193910. Keeping this open would only cost review attention for an argument I no longer make.
